recipe_app/views.py

```diff
@@ -0,0 +1,179 @@
+        specific `id`. If `id` is not provided, the function will retrieve all objects that match the query
+        parameter `name`
+
+        Returns:
+          The code is returning a JsonResponse object. If no `id` is passed, it returns a
+        JsonResponse containing a list of all recipes. If the `id`parameter is passed, it
+        returns a JsonResponse containing a single Recipe object specified by `id`.
+        """
+        if id is None:
+            return JsonResponse(
+                RecipeSerializer(
+                    self._handle_get_all(query_params=request.GET.dict()), many=True
+                ).data,
+                safe=False,
+            )
+
+        return JsonResponse(
+            RecipeSerializer(self._handle_get_by_id(id)).data, safe=False
+        )
+
+    @transaction.atomic
+    def post(self, request: HttpRequest) -> JsonResponse:
+        """
+        Function that handles POST requests, for creating a new recipe and its associated ingredients based
+        on the request body.
+
+        Args:
+          request (HttpRequest): The `request` parameter is an instance of the `HttpRequest` class, which
+        represents an HTTP request made to the server. It contains information about the request, such as
+        the HTTP method (GET, POST, etc.), headers, and body.
+
+        Returns:
+          JsonResponse object of the created Recipe object.
+        """
+        body = JSONParser().parse(request)
+        serializer = RecipeSerializer(data=body)
+
+        if serializer.is_valid():
+            serializer.save()
+            return JsonResponse(serializer.data, status=201)
+
+        return JsonResponse(serializer.errors, status=400)
+
+    def patch(self, request: HttpRequest, id: int) -> JsonResponse:
+        """
+        Function to handle PATCH requests. The function updates a recipe object with new data provided in the request body and returns the
+        updated recipe as a JSON response.
+
+        Args:
+          request (HttpRequest): An instance of the HttpRequest class, which represents an HTTP request made
+        by a client.
+          id (int): The `id` parameter is an integer that represents the unique identifier of a recipe. It
+        is used to retrieve the recipe from the database and update its properties.
+
+        Returns:
+          a JsonResponse object with the updated Recipe object.
+        """
+        recipe = self._handle_get_by_id(id)
+        body = json.loads(request.body.decode("utf-8"))
+
+        if "name" in body:
+            recipe.name = body["name"]
+
+        if "description" in body:
+            recipe.description = body["description"]
+
+        if "icon" in body:
+            if emoji.is_emoji(body["icon"]):
+                recipe.icon = body["icon"]
+            else:
+                return JsonResponse({"message": "Invalid emoji for icon"}, status=400)
```

Comment:
Don't we already do this in the model field? I believe the model serializer should take care of this for us: https://www.django-rest-framework.org/api-guide/serializers/#modelserializer

> It will automatically generate validators for the serializer, such as unique_together validators.

Reply:
The `recipe.save` doesn't utilise the validators by default, meaning when removing this the `test_modify_recipe_icon_invalid` test fails. However, I have made this utilise the serialisers by adding a `recipe.full_clean()` call inside of a new `_validate_model` function, which validates the entire model as expected.